proposals/0569-multiline-strings.rst

```diff
@@ -0,0 +1,629 @@
+
+      * Ignore any characters preceding the first newline
+      * Blank lines and lines with only whitespace should not be included in this calculation
+
+   #. Remove exactly one newline from the beginning of the string (if one exists)
+
+   #. Interpret escape sequences (occurs after removing whitespace prefix so that literal ``\n`` characters are not included)
+
+#. After parsing, it becomes indistinguishable to the equivalent single-quoted string (modulo annotations for exact-printing)
+
+Line terminators will match whatever line terminators the user is using in the file.
+
+Examples
+--------
+
+Line continuations
+~~~~~~~~~~~~~~~~~~
+
+Line continuations are collapsed first and not included in the whitespace calculation
+
+::
+
+  s =
+      """
+        a b\
+    \ c d e
+        f g
+      """
+
+  -- equivalent to
+  s' = "a b c d e\nf g\n"
+
+Remove common whitespace
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+In the below examples, leading spaces will be marked as ``.`` for visibility.
+
+::
+
+  s =
+        """
+  ......a b c
+
+  ......d e f
+  ..
+  ....g h i
+  ......"""
+
+  -- equivalent to
+  s' = "..a b c\n\n..d e f\n\ng h i\n.."
+
+After lexing, the initial multiline above is parsed as
+
+::
+
+  [ "......a.b.c"
+  , ""
+  , "......d.e.f"
+  , ".."
+  , "....g.h.i"
+  , "......"
+  ]
+
+The blank line + the whitespace-only line are excluded from the calculation, and we calculate 4 spaces as the shared whitespace prefix, which are removed from every line.
+
+Note that the whitespace preceding the closing ``"""`` is included. This implies that there will be a trailing newline (see the "Trailing newline" example for more information).
+
+Ignore leading characters
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The common prefix calculation ignores all characters preceding the first newline. This means that characters immediately after the ``"""`` delimiter will be included verbatim. The same would occur with a line continuation (since line continuations are collapsed before the prefix calculation).
+
+::
+
+  s =
+    """Line 1
+       Line 2
+    Line 3
+    """
+
+  s_2 =
+    """\
+   \Line 1
+       Line 2
+    Line 3
+    """
+
+  -- equivalent to
+  s' = "Line 1\n   Line 2\nLine 3"
+
+This implies that normal strings could also be written using ``"""``
+
+::
+
+  -- the following are equivalent
+  s = """hello world"""
+  s' = "hello world"
+
+Mixing tabs and spaces
+~~~~~~~~~~~~~~~~~~~~~~
+
+Per the `Haskell report <https://www.haskell.org/onlinereport/haskell2010/haskellch10.html#x17-17800010.3>`_, tabs will be treated as 8 spaces in the calculation. If the common prefix is in the middle of a tab (which would only happen if mixing spaces and tabs), the tab will be converted to 8 spaces first. But don't do this.
+
+In the following example, two ``-`` characters will represent 1 tab, and 1 ``.`` character will represent 4 spaces.
+
+::
+
+  s = """
+  ------a
+  ....b
+  .------c
+  ...."""
+
+  -- equivalent to
+  s' = "--a\nb\n.--c\n"
```

Comment:
Honestly, this is a very difficult example to parse.

Reply:
I don't disagree. Any ideas on how to improve it? Specifically, I wanted to demonstrate what happens when the common whitespace removal occurs in the middle of a tab character.

Maybe if I just said:

> e.g. say a multiline string contains 3 lines:
>
> * Line 1 - 2 tabs before content
> * Line 2 - 8 spaces before content
> * Line 3 - 2 spaces + 2 tabs before content
>
> The resulting string would have the following lines:
>
> * Line 1 - 1 tab before content
> * Line 2 - no spaces before content
> * Line 3 - 2 spaces + 1 tab before content

Would that be better?